core/engine.go

```diff
@@ -106,128 +106,182 @@ func NewEngine(ex lib.ExecutionScheduler, o lib.Options, logger *logrus.Logger)
 	return e, nil
 }
 
-// Init is used to initialize the execution scheduler. That's a costly operation, since it
-// initializes all of the planned VUs and could potentially take a long time.
-func (e *Engine) Init(ctx context.Context) error {
-	return e.ExecutionScheduler.Init(ctx, e.Samples)
-}
-
-func (e *Engine) setRunStatus(status lib.RunStatus) {
-	for _, c := range e.Collectors {
-		c.SetRunStatus(status)
+// Init is used to initialize the execution scheduler and all metrics processing
+// in the engine. The first is a costly operation, since it initializes all of
+// the planned VUs and could potentially take a long time. It either returns an
+// error immediately, or it returns test run() and wait() functions.
+//
+// Things to note:
+//  - The first lambda, Run(), synchronously executes the actual load test.
+//  - It can be prematurely aborted by cancelling the runCtx - this won't stop
+//    the metrics collection by the Engine.
+//  - Stopping the metrics collection can be done at any time after Run() has
+//    returned by cancelling the globalCtx
+//  - The second returned lambda can be used to wait for that process to finish.
+func (e *Engine) Init(globalCtx, runCtx context.Context) (run func() error, wait func(), err error) {
+	e.logger.Debug("Initialization starting...")
+	// TODO: if we ever need metrics processing in the init context, we can move
+	// this below the other components... or even start them concurrently?
+	if err := e.ExecutionScheduler.Init(runCtx, e.Samples); err != nil {
+		return nil, nil, err
 	}
-}
 
-func (e *Engine) Run(ctx context.Context) error {
-	e.runLock.Lock()
-	defer e.runLock.Unlock()
+	// TODO: move all of this in a separate struct? see main TODO above
+
+	runSubCtx, runSubCancel := context.WithCancel(runCtx)
 
-	e.logger.Debug("Engine: Starting with parameters...")
+	resultCh := make(chan error)
+	runFn := func() error {
+		e.logger.Debug("Execution scheduler starting...")
+		err := e.ExecutionScheduler.Run(globalCtx, runSubCtx, e.Samples)
+		e.logger.WithError(err).Debug("Execution scheduler terminated")
 
-	collectorwg := sync.WaitGroup{}
-	collectorctx, collectorcancel := context.WithCancel(context.Background())
+		select {
+		case <-runSubCtx.Done():
+			// do nothing, the test run was aborted somehow
+		default:
+			resultCh <- err // we finished normally, so send the result
+		}
 
+		return err
+	}
+	waitFn := e.startBackgroundProcesses(globalCtx, runCtx, resultCh, runSubCancel)
+	return runFn, waitFn, nil
+}
+
+// This starts a bunch of goroutines to process metrics, thresholds, and set the
+// test run status when it ends. It returns a function that can be used after
+// the provided context is called, to wait for the complete winding down of all
+// started goroutines.
+func (e *Engine) startBackgroundProcesses( //nolint:funlen
```

Comment:
I would prefer splitting this method into individual `start*` methods for each process. Right now this is a dumping ground for any background process handling, and having them split would be useful for testing and scenarios where you don't need to initialize all of them. Not a big deal, though.

Reply:
These background processes have inter-dependencies between them... See `thresholdAbortChan`. They also have to deal with the same `WaitGroup`, so this seemed cleaner - some of the code is in separate methods `see `processMetrics()`, `runMetricsEmission()`, `processThresholds()`), but the starting and synchronization of all of these processes is in `startBackgroundProcesses()`